llms/mlx_lm/models/gpt_bigcode.py

```diff
@@ -0,0 +1,211 @@
+
+        self.scale = head_dim**-0.5
+
+        if hasattr(args, "attention_bias"):
+            attention_bias = args.attention_bias
+        else:
+            attention_bias = False
+
+        self.c_attn = nn.Linear(dim, dim + 2 * self.kv_dim, bias=attention_bias)
+        self.c_proj = nn.Linear(dim, dim, bias=attention_bias)
+        self.resid_dropout = nn.Dropout(args.resid_pdrop)
+
+    def __call__(
+        self,
+        x: mx.array,
+        mask: Optional[mx.array] = None,
+        cache: Optional[Tuple[mx.array, mx.array]] = None,
+    ) -> mx.array:
+        B, L, D = x.shape
+
+        qkv = self.c_attn(x)
+        queries, keys, values = mx.split(
+            qkv, [self.dim, self.dim + self.kv_dim], axis=-1
+        )
+
+        # Prepare the queries, keys and values for the attention computation
+        queries = queries.reshape(B, L, self.n_heads, -1).transpose(0, 2, 1, 3)
+        keys = keys.reshape(B, L, self.n_kv_heads, -1).transpose(0, 2, 1, 3)
+        values = values.reshape(B, L, self.n_kv_heads, -1).transpose(0, 2, 1, 3)
+
+        if cache is not None:
+            keys, values = cache.update_and_fetch(keys, values)
+
+        output = mx.fast.scaled_dot_product_attention(
+            queries, keys, values, scale=self.scale, mask=mask
+        )
+        output = output.transpose(0, 2, 1, 3).reshape(B, L, -1)
+        output = self.c_proj(output)
+        return self.resid_dropout(output)
+
+
+class MLP(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+
+        dim = args.n_embd
+        hidden_dim = args.n_inner
+        if hasattr(args, "mlp_bias"):
+            mlp_bias = args.mlp_bias
+        else:
+            mlp_bias = False
+
+        self.c_fc = nn.Linear(dim, hidden_dim, bias=mlp_bias)
+        self.c_proj = nn.Linear(hidden_dim, dim, bias=mlp_bias)
+        self.dropout = nn.Dropout(args.resid_pdrop)
+
+    def __call__(self, x) -> mx.array:
+        return self.dropout(self.c_proj(nn.gelu(self.c_fc(x))))
+
+
+class TransformerBlock(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+        self.n_head = args.n_head
+        self.n_embd = args.n_embd
+        self.attn = Attention(args)
+        self.mlp = MLP(args)
+        self.ln_1 = nn.LayerNorm(args.n_embd, eps=args.layer_norm_epsilon)
+        self.ln_2 = nn.LayerNorm(args.n_embd, eps=args.layer_norm_epsilon)
+        self.args = args
+
+    def __call__(
+        self,
+        x: mx.array,
+        mask: Optional[mx.array] = None,
+        cache: Optional[Tuple[mx.array, mx.array]] = None,
+    ) -> mx.array:
+        r = self.attn(self.ln_1(x), mask, cache)
+        h = x + r
+        r = self.mlp(self.ln_2(h))
+        out = h + r
+        return out
+
+
+def create_additive_causal_mask(N: int, offset: int = 0):
+    rinds = mx.arange(offset + N)
+    linds = mx.arange(offset, offset + N) if offset else rinds
+    mask = linds[:, None] < rinds[None]
+    return mask * -1e9
```

Comment:
would you mind refactoring this into `base.py` and calling that from this model and the `llama.py` model? Ultimately we'd like all the models to use this same version but we can start with that and then update from there rather than duplicate the function everywhere.

Reply:
Sure, it's done ✅